### ISSUE TITLE
Is the project still alive ?

### ISSUE THREAD

**Issue by bric3** [nokeedev/gradle-native#855]
There's no commit since may 2023
<img width="893" alt="image" src="https://github.com/nokeedev/gradle-native/assets/803621/27e0ec82-369b-4035-84d5-75c56fd19297">

I would understand since most work has been done by @lacasseio, but that is not good news for the rest of te users. The current build don't run on recent versions of Gradle.

**Comment by lacasseio** [comment]
Nothing to worry about. I'm still working on the project. I was distracted with the Toolbox project for the last two months as I'm trying to make Nokee build more maintainable. I'm aware of the issue with the recent Gradle build and that spin up more decisions in terms of the direction to take. Gradle is unfortunately not too communicative about their long term internal decision which broke a couple things on my side. Despite voicing my concerns on lots of those breakage, the newer members seem to move in very different directions then what the "essence" of Gradle is so I need to adapt Nokee.

I also took some time off early in the summer. I'm hoping to get things rolling again very soon.

**Comment by bric3** [comment]
Thank you for the update, I understand you being busy elsewhere, that's what open source is.

Indeed these API changes are hard to handle. And rarely explained.

Anyway take care and thank you for the hard work already done.

**Comment by bric3** [comment]
Just spotted this https://github.com/nokeedev/gradle-native/pull/858 (major work toward a different 1.0)

**Comment by bric3** [comment]
Hi there, sorry to be nagging, but is there any news ?
Maybe help ?

**Comment by 12345ieee** [comment]
Can we please have some news?
Even if the news are "project will not be updated anymore and #858 will not be finished", so that users can start finding an alternative.

**Comment by lacasseio** [comment]
Sorry for the longer-than-expected hiatus. I'm still working on the project. My time has been quite heavily dedicated to clients at the moment so the effort has shifted toward samples which I'm hoping to release something about soon.

@bric3 @12345ieee Just to plan a little bit better, are you JNI users or more C/C++ users?

**Comment by bric3** [comment]
@lacasseio A bit of pure native, but clearly more JNI / FFM

**Comment by 12345ieee** [comment]
JNI here as well.

We know how open source maintainership is, don't feel pressured, my main issue is that this is the last blocker to migrate to java 21 from 17.

**Comment by lacasseio** [comment]
Don't worry, I hope to focus a bit more next month. I'm in the middle of trying to get a sample portal release. With my clients, I built a lot of samples and they are a bit distributed everywhere but also some of the work cross-cut with Nokee and the current core native support. Hold tight, I'm still working on Gradle native every day even if this repo doesn't reflect that.

**Comment by bric3** [comment]
@lacasseio Thank you for the feedback, that is much appreciated !

**Comment by 12345ieee** [comment]
I'll stop bothering you after this, is there any hope for something to move inside a month or so?

**Comment by bric3** [comment]
@lacasseio is the nokee-companion the new thing ?

**Comment by lacasseio** [comment]
Yes and no. I started the Gradle Native effort after leaving Gradle becuase there was no investment in the native ecosystem. Despite my lack of update, everyday I'm working/thinking/planning within the native ecosystem for Gradle build tool. The original effort (this repo) was a direct follow up to the newer native plugins. Despite how "clean" the newer plugin feels, they got a lot wrong which I slowly realized. This is not to say that the project is abandoned. To the contrary, I have been reflecting more deeply in how best to proceed. Pair this with some very strange direction change within Gradle build tool (ex. deprecation of very useful APIs was the biggest hit for my work).

This repo focused on a top down approach (starts with the DSL and move toward the details). With things like the declarative DSL, it's obvious that approach is not a good one. The Gradle team is doing a great job in that effort but it has very little consideration on complex ecosystems like the native ecosystem. For this reason, I shifted toward a bottom up approach. The native-companion is a representation of that effort specifically for clients using the newer native plugin (and anyone in the same situation). The shift was to alleviate immediate pain following the bottom down approach. In fact most fixes provided by the native-companion keeps the DSL as intact as possible while providing fixes that the Gradle team are unable to provide.

That being said, I completely understand the feeling of abandonment around the native but rest assured I'm still 100% invested in the native ecosystem under Gradle. Now, what does the bottom up approach look like in practice and for the future of Gradle Native?

The conclusion I came with much reflection is a strict DSL matter very little for complex ecosystems like native. The most important rule of thumb in Gradle is ensuring the configuration complexity follows a square root graph (as opposed to exponential). For this to happen, we need a way to make the complexity of "a native thing" grow organically. This is what I'm calling an accordion configuration. A "native thing" (C++ source set  header set, library, binary, etc.) may have different representations (such as but not limited to the DSL used). However, "the native thing" has a meaning. That meaning doesn't change but the representation may change. Let's take for example the include roots. They can be represented as a lazy `FileCollection`. But they could be represented as an eager `List<File>` or by anything that would "mean" a collection of directory, like a, somewhat hypothetical, `HeaderSet`. The complexity comes when we want to give more meaning to the include roots, like the files to consider as the actual valid headers under each directories (known as the filter) for publishing the headers, or some implicit compiler arguments to use when compiling against those headers, or the dependencies that are true only for that set of headers. All those levels of complexity are valid under different scenarios. Not everyone requires that complexity, but, when you need them, it should be as simple as "upgrading" the representation. The current native plugins (and mine) didn't consider this possibility. This is not much of a new problem. The Gradle team tried solving this "idiomatic" problem several times but failed every time. The lack of progress on their latest design document about the subject shows they are hitting the same issue that I have been pondering my head on for the last couple years. However, I'm pretty confident I found a way forward and will be starting more active work.

What's the timeline? I can't say. My first priority is my client's needs. If you need something specific, we should talk and find a path forward for you. My second priority is to unblock the native ecosystem progression. I hope this answers some questions.

**Comment by 12345ieee** [comment]
I have since moved to the java 22+ FFM API.

I found it quite nicer than JNI, and I hope that is the step forward for java.

Good luck on your efforts to get gradle to work for a native build, thanks for this tool I used for many years.

**Comment by lacasseio** [comment]
Thanks for the feedback. I was about to add a small note specifically about JNI. My JNI plugin (and the other in circulation) are good-ish demonstration on various ways to solve the same issue. Where all of them fails (mine included) is the focus on a stricter DSL. The strict modeling prevent any additional composition (or recomposition) to offer different or alternative functionality. As a quite demonstration, I was able to [write a super basic, but functional, JNI plugin for a sample I wrote for multiplatform publishing](https://github.com/nokeedev/nokee-companion/blob/main/samples%2Fcpp-additional-publishing-variants%2FREADME.md). There nothing inherently wrong with that implementation either. The key concept I'm trying to focus is on "the meaning of each things" so that compoment that have similar meaning can just understand each other. Part of that work is to allow ways to pluck out specific value out of a property in a way to override some rule or chain rule. Essentially giving more flexibility to the build/plugin authors. Of course, that flexibility come at a cost but it should never be exponential. As long as we can declare the meaning of what is part of a component, the representation is secondary. In some ways, you could declare your own representation if you dont like the one provided focusing only on the things you care about.

I know it sounds a lot like I'm waving my hand in the air in a very abstract way. The key point to remember is moving towards a meaningful model as opposed to a "declarative", "software", or whatever things that were tried for representing what is being built.

**Comment by bric3** [comment]
Thank you @lacasseio for this update 🙇

In my opinion, these are very sensitive ideas and strategy. And I believe they ought to be more visible e.g. in the nokee blog, rather than only mentioned in this issue.
